### ISSUE TITLE
add multilib tests

### ISSUE THREAD

**Issue by stsp** [dosemu2/dosemu2#1387]
Would be cool to have multilib tests
added to weekly test suit.
@stefand reminded me that ubuntu
isn't dropping it at all, I was confused
saying it does.

Andrew, would you like to add such tests?
Would also be cool to add pmdapi test
to weekly too.

**Comment by stsp** [comment]
Also this:
https://github.com/dosemu2/dosemu2/pull/1397#issuecomment-774787166
says there are no tests for rw open of
ro files.

Another wish was to add pmdapi as
a test-case.

**Comment by andrewbird** [comment]
> says there are no tests for rw open of
ro files.

DOS r/w open of files with DOS r/o attribute set, or Unix r/o file?

**Comment by stsp** [comment]
I really can't remember.
Do we have either?
Anyway, even talking about a host r/o,
I wouldn't expect eg mscdex to allow an
rw open on a cdrom.

**Comment by andrewbird** [comment]
Does DOS have a concept of read-only device ( except floppies), or are CDROM files just presented as files with r/o attribute set?

**Comment by andrewbird** [comment]
> Do we have either?
 
Definitely not r/o unix file, not sure about r/o DOS attribute, will need to check.

**Comment by stsp** [comment]
> Does DOS have a concept of read-only device ( except floppies)

If it has them for floppies, that should
be enough. :)

> or are CDROM files just presented as files with r/o attribute set?

Only tests can tell. We support cdrom and
mscdex can be used. Unfortunately not on
images, but only on a real cdrom. :(

**Comment by stsp** [comment]
With floppy I believe it just raises the
critical error on write so its probably
the most uninteresting case.

**Comment by stefand** [comment]
> I really can't remember.
> Do we have either?
> Anyway, even talking about a host r/o,
> I wouldn't expect eg mscdex to allow an
> rw open on a cdrom.

Bug 989?

**Comment by stsp** [comment]
Yes, its #989 thanks.
So it states clearly that cdrom doesn't
force the R/O attr and doesn't prevent
an RW open.
So if we are to add any tests, they should
test the open of R/O (dos attr) files, and
should make sure that lredir -C or lredir -R
produces the files that can be opened RW
(but cannot be written to).

**Comment by stsp** [comment]
Andrew, what do you think about
starting tests in parallel? Something
like a 10x speed-up?

**Comment by stuaxo** [comment]
This exists, which might help:
https://pypi.org/project/unittest-parallel/

Some tests need some help to be parallelizable, it wouldn't surprise me if DOSEMU2s fall into this category, but it's worth a go as a first step.


I wonder how much time is spent compiling assembly for tests in this?  Implementing a cache might not be too much work and could help.

Something like storing the output in directories named as the sha256 of the content to be compiled, + later checking if they exist and copying the files, instead of compiling.

**Comment by stsp** [comment]
So is multilib still viable in the
near future? I guess ubuntu is
not dropping it so why not to
have the 32bit CI builds?

**Comment by stefand** [comment]
I think multilib is here to stay, I believe Valve stated publicly that they aren't going to port their old catalogue to 64 bit. So while Steam may ship higher level libs themselves they'll always need a 32 bit loader and 32 bit support in the kernel.

Though the question for Dosemu is, what for? Is there something special about running as a 32 bit process on a 64 bit kernel that 64 bit process on 64 bit kernel doesn't have? I think what's more relevant is 32 bit process on 32 bit kernel (and 32 bit CPU, i.e. keep working on old HW). That's not multilib, but if github's test runners don't offer this setup then maybe 32 bit process on 64 bit kernel might be the best possible substitute.

Other than testing some corner cases I don't see why I would run a 32 bit dosemu on my 64 bit system.

**Comment by stsp** [comment]
You would never want to run this, no.
The only reason is to keep testing the
32bit builds.

**Comment by stsp** [comment]
> That's not multilib, but if github's test runners don't offer this setup then maybe 32 bit process
> on 64 bit kernel might be the best possible substitute.

Isn't this multilib?

**Comment by stefand** [comment]
> > That's not multilib, but if github's test runners don't offer this setup then maybe 32 bit process
> > on 64 bit kernel might be the best possible substitute.
> 
> Isn't this multilib?

Yes it is multilib, but it is only a limited substitute for a pure 32 bit setup, which is what you are actually interested in.

The most obvious differences between 32-on-64 multilib and pure 32 bit that come to my mind are vm86 and address space. A 32 bit process on a 64 bit kernel has all 4 GB available. A 32 bit process on a 32 bit kernel usually has to deal with a 3-1 or 2-2 memsplit.

**Comment by stsp** [comment]
Its enough to test the build though.
We dont even have to run (many) tests
on that.
